Tasks/JenkinsDownloadArtifacts/ArtifactDetails/JenkinsRestClient.ts

```diff
@@ -141,4 +212,127 @@ export class JenkinsRestClient {
+        const jenkinsBuild: string = tl.getInput('jenkinsBuild', true);
+        if (jenkinsBuild === 'LastSuccessfulBuild') {
+            return this.GetLastSuccessfulBuild();
+        } else {
+            let defer = Q.defer<JenkinsJobDetails>();
+            const jobName: string = tl.getInput('jobName', true);
+            let jobType: string = "";
+            let buildIdStr: string = tl.getInput('jenkinsBuildNumber');
+            let branchName: string;
+
+            this.GetJobType().then((result: string) => {
+                jobType = result;
+            }).fin(() => {
+
+                let isMultibranchPipeline = jobType.toLowerCase() === JenkinsJobTypes.MultibranchPipeline.toLowerCase();
+
+                tl.debug(`Found Jenkins Job type ${jobType} and isMultibranchPipeline ${isMultibranchPipeline}`);
+                // if its multibranch pipeline extract the branch name and buildId from the buildNumber input
+                if (isMultibranchPipeline && !!buildIdStr && buildIdStr.indexOf(JenkinsRestClient.JenkinsFolderSeperator) != -1) {
+                    tl.debug(`Extracting branchname and buildId from selected version`);
+                    let position: number = buildIdStr.indexOf(JenkinsRestClient.JenkinsFolderSeperator);
+                    branchName = buildIdStr.substring(0, position);
+                    buildIdStr = buildIdStr.substring(position + 1);
+                }
+
+                const buildId = parseInt(buildIdStr);
+                if (isNaN(buildId)) {
+                    defer.reject(new Error(tl.loc("InvalidBuildId", buildIdStr)));
+                }
+                else {
+                    console.log(tl.loc("FoundJenkinsJobDetails", jobName, buildId, branchName));
+                    defer.resolve(new JenkinsJobDetails(jobName, buildId, jobType, branchName));
+                }
+            });
+
+            return defer.promise;
+        }
+    }
+
+    public GetLastSuccessfulBuild(): Q.Promise<JenkinsJobDetails> {
+        let defer = Q.defer<JenkinsJobDetails>();
+        let jobType: string = "";
+        const jobName: string = tl.getInput('jobName', true);
+
+        this.GetJobType().then((result: string) => {
+            jobType = result;
+        }).fin(() => {
+
+            let jenkinsTreeParameter = "lastSuccessfulBuild[id,displayname]";
+            let handlerbarSource = "{{lastSuccessfulBuild.id}}";
+            const isMultibranchPipeline = jobType.toLowerCase() === JenkinsJobTypes.MultibranchPipeline.toLowerCase();
+
+            tl.debug(`Found Jenkins Job type ${jobType} and isMultibranchPipeline ${isMultibranchPipeline}`);
+            if (isMultibranchPipeline) {
+                jenkinsTreeParameter = "jobs[name,lastSuccessfulBuild[id,displayName,timestamp]]"
+                handlerbarSource = "{{#with (selectMaxOf jobs 'lastSuccessfulBuild.timestamp') as |job|}}{ \"branchName\": \"{{job.name}}\", \"buildId\": \"{{job.lastSuccessfulBuild.id}}\" }{{/with}}"
```

Comment:
do we have #with in the result templates?

Reply:
We have #with in our implementation. But we don't have support for subexpression. We can only have {{#with something}}.. We can't  have {{#with (something)}}... 

Here i need this #with only because the javascript implementation of block helper returns me a string. ie., the output of #selectMaxOf is a string. But i have a subexpression to execute on top of the result. Using #with on top of #selectMaxOf converts the string to JObject and I can operation the subexpression on it.

In our implementation on the service with result template, we treat them as object. Hence I don't need this #with helper at all. In our extension the same thing you will see with out the #with helper.